File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int start = 1;
        int end = *max_element(piles.begin(),piles.end());
        int ans = 0;
        while(start<=end){
            int mid = start + (end-start)/2;
            if(caneat(mid,piles,h)){
                end = mid-1;
                ans = mid;
            }else start = mid+1;
        }
        return ans;
    }

    bool caneat(int k,vector<int>& piles,int h){
        long long  sum = 0;
        for(int num : piles){
            sum+=(num+k-1)/k;
        }
        return sum<=h;
    };
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    boolean canEat(int[] piles, int h, int speed) {
        int total = 0;
        for (int pile : piles) {
            total += (pile + speed - 1) / speed;
            if (total > h) return false;
        }
        return true;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int low = 1;
        int high = Arrays.stream(piles).max().getAsInt();

        while(low < high){
            int mid = low + (high - low)/2;
            if(canEat(piles,h,mid)){
                high = mid;
            }else{
                low = mid+1;
            } 
        }
        return low;   
    }
}